// @huddle01/server-sdk/dist/chunk-2ZBCNKEG.js
// package.json
var package_default = {
  name: "@huddle01/server-sdk",
  version: "2.5.7",
  author: "OmGuptaIND",
  private: false,
  description: "The Huddle01 Server SDK allows you to perform protected admin actions on your server side, like generating peer access tokens and starting and stopping meeting recordings and livestreams.",
  main: "dist/index.js",
  types: "dist/index.d.ts",
  type: "module",
  exports: {
    "./webhooks": {
      import: "./dist/webhooks/index.js",
      require: "./dist/webhooks/index.cjs",
      types: "./dist/webhooks/index.d.ts"
    },
    "./auth": {
      import: "./dist/auth/index.js",
      require: "./dist/auth/index.cjs",
      types: "./dist/auth/index.d.ts"
    },
    "./recorder": {
      import: "./dist/recorder/index.js",
      require: "./dist/recorder/index.cjs",
      types: "./dist/recorder/index.d.ts"
    },
    "./api": {
      import: "./dist/api/index.js",
      require: "./dist/api/index.cjs",
      types: "./dist/api/index.d.ts"
    }
  },
  license: "MIT",
  scripts: {
    clean: "rm -rf .turbo node_modules",
    build: "rm -rf dist && tsup",
    pub: "pnpm run build && pnpm publish --access public --no-git-checks",
    "pub:beta": "pnpm run build && pnpm publish --access public --tag beta --no-git-checks",
    "pub:alpha": "pnpm run build && pnpm publish --access public --tag alpha --no-git-checks",
    "pub:dev": "pnpm run build && pnpm publish --access public --tag dev --no-git-checks",
    typecheck: "tsc --noEmit"
  },
  keywords: [
    "huddle01",
    "server",
    "sdk",
    "recorder",
    "livestream",
    "webhooks",
    "api"
  ],
  files: [
    "dist/*",
    "index.d.ts",
    "auth.d.ts",
    "recorder.d.ts",
    "webhooks.d.ts",
    "api.d.ts"
  ],
  devDependencies: {
    "@infra/api": "workspace:*",
    "@shoshin/tsconfig": "workspace:^0.1.0",
    tsup: "^8.3.0",
    typescript: "^5.6.2"
  },
  dependencies: {
    hono: "^4.7.0",
    jose: "^5.9.3",
    zod: "^3.24.1"
  }
};

// src/constants/index.ts
var MAX_METADATA_SIZE = 5 * 1024;
var MAX_ROOM_METADATA_SIZE = 1024;
var INFRA_URL = "https://infra-api.huddle01.workers.dev";
var SDK_VERSION = package_default.version;

export { INFRA_URL, MAX_METADATA_SIZE, MAX_ROOM_METADATA_SIZE, SDK_VERSION };
